src/c_kzg_4844.c

```diff
@@ -2378,231 +2379,84 @@ typedef struct {
+    for (size_t i = 1; i < roots_len; i++) {
+        blst_fr_cneg(&neg_root, &roots[i], true);
 
-    ret = fft_fr(mul_eval_ps, p_padded, len_out, s);
-    if (ret != C_KZG_OK) goto out;
+        poly[i] = neg_root;
+        blst_fr_add(&poly[i], &poly[i], &poly[i - 1]);
 
-    for (size_t i = 0; i < partial_count - 1; i++) {
-        ret = pad_p(
-            p_padded, partials[i].length, partials[i].coeffs, partials[i].length
-        );
-        if (ret != C_KZG_OK) goto out;
-        ret = fft_fr(p_eval, p_padded, len_out, s);
-        if (ret != C_KZG_OK) goto out;
-        for (size_t j = 0; j < len_out; j++) {
-            blst_fr_mul(&mul_eval_ps[j], &mul_eval_ps[j], &p_eval[j]);
+        for (size_t j = i - 1; j > 0; j--) {
+            blst_fr_mul(&poly[j], &poly[j], &neg_root);
+            blst_fr_add(&poly[j], &poly[j], &poly[j - 1]);
         }
+        blst_fr_mul(&poly[0], &poly[0], &neg_root);
     }
 
-    ret = ifft_fr(out->coeffs, mul_eval_ps, len_out, s);
-    if (ret != C_KZG_OK) goto out;
-
-    out->length = out_degree + 1;
+    poly[roots_len] = FR_ONE;
+    *poly_len = roots_len + 1;
 
-out:
-    return ret;
+    return C_KZG_OK;
 }
 
-/**
- * Calculate the minimal polynomial that evaluates to zero for the powers of
- * roots of unity that correspond to missing indices.
+/* Computes the minimal polynomial that evaluates to zero at equally
+ * spaced chosen roots of unity in the domain of size `FIELD_ELEMENTS_PER_BLOB`.
  *
- * This is done simply by multiplying together `(x - r^i)` for all the `i` that
- * are missing indices, using a combination of direct multiplication
- * (#do_zero_poly_mul_partial) and iterated multiplication via convolution
- * (#reduce_partials).
+ * The roots of unity are chosen based on the missing cell indices. If the i'th
+ * cell is missing, then the i'th root of unity will be zero on the polynomial
+ * computed, along with every `CELLS_PER_EXT_BLOB` spaced root of unity in the
+ * domain.
  *
- * @param[out]  zero_poly       The zero polynomial
- * @param[out]  zero_poly_len   The zero polynomial length
- * @param[in]   missing_indices The indices of the missing coefficients
- * @param[in]   len_missing     The number of missing indices
- * @param[in]   s               The trusted setup
+ * @param[in,out]   zero_poly      The zero polynomial for cell indices
+ * @param[in,out]   zero_poly_len  The length of zero_poly
+ * @param[in]       missing_cell_indices     The array of missing cell indices
+ * @param[in]       len_missing_cells The number of missing cell indices
+ * @param[in]       s         The trusted setup
  *
- * @remark This does not work if all indices are missing.
- * @remark Unused coefficients are set to zero.
+ * @remark This does not work if all cells are missing.
+ * @remark `missing_cell_indices` are assumed to be less than
+ * `CELLS_PER_EXT_BLOB`.
  */
-static C_KZG_RET zero_polynomial_via_multiplication(
+static C_KZG_RET zero_polynomial_periodic_roots(
```

Comment:
unsure about the postfix "periodic_roots" -- happy to change this to anything else that is clearer upon first read